refactor(relay): move stream-handling away from `{In,Out}boundUpgrade`

dgarus opened the pull request:
## Description

<!--
Please write a summary of your changes and why you made them.
This section will appear as the commit message after merging.
Please craft it accordingly.
For a quick primer on good commit mesages, check out this blog post: https://cbea.ms/git-commit/

Please include any relevant issues in here, for example:

Related https://github.com/libp2p/rust-libp2p/issues/ABCD.
Fixes https://github.com/libp2p/rust-libp2p/issues/XYZ.
-->

Fixes: #4075.

## Notes & open questions

<!--
Any notes, remarks or open questions you have to make about the PR which don't need to go into the final commit message.
-->

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

dgarus committed:
Refactoring of stop protocol

dgarus committed:
Merge branch 'master' into 4075-relay

thomaseizinger commented on protocols/relay/src/behaviour/handler.rs:
```diff
@@ -716,6 +741,21 @@ impl ConnectionHandler for Handler {
+                    Either::Right(e),
+                ))),
+            };
+        }
```
I think we can slightly simplify this:

Instead of returning the oneshot from the future, we can construct the oneshot early within `send_stop_message_and_process_result` and directly push it into `self.alive_lend_out_substreams`.

If the outbound-stop-future fails, the other end of the oneshot will be dropped and we will automatically clean this up. Whilst we are negotiating the outbound stream, it is still alive so I don't think there is any harm in constructing it early.

dgarus replied:
> I think we can slightly simplify this:
> 
> Instead of returning the oneshot from the future, we can construct the oneshot early within `send_stop_message_and_process_result` and directly push it into `self.alive_lend_out_substreams`.
> 
> If the outbound-stop-future fails, the other end of the oneshot will be dropped and we will automatically clean this up. Whilst we are negotiating the outbound stream, it is still alive so I don't think there is any harm in constructing it early.

When I try to change `self.alive_lend_out_substreams` inside the future, i get the following:
```
error: future cannot be sent between threads safely
   --> protocols/relay/src/behaviour/handler.rs:440:9
    |
440 | /         async move {
441 | |             substream.send(msg).await?;
442 | |
443 | |             let proto::StopMessage {
...   |
515 | |             })
516 | |         }
    | |_________^ future created by async block is not `Send`
    |
    = help: the trait `Sync` is not implemented for `(dyn futures::Future<Output = Result<behaviour::handler::Event, outbound_stop::FatalUpgradeError>> + std::marker::Send + 'static)`
note: captured value is not `Send` because `&` references cannot be sent unless their referent is `Sync`
   --> protocols/relay/src/behaviour/handler.rs:506:13
    |
506 |             self.alive_lend_out_substreams.push(rx);
    |             ^^^^ has type `&behaviour::handler::Handler` which is not `Send`, because `behaviour::handler::Handler` is not `Sync`
note: required by a bound in `futures::FutureExt::boxed`
   --> /Users/denisgarus/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.28/src/future/future/mod.rs:520:23
    |
520 |         Self: Sized + Send + 'a,
    |                       ^^^^ required by this bound in `FutureExt::boxed`
 ```
That is why I return the `oneshot` from the future.
Could you please tell me the right way to fix this?

thomaseizinger replied:
You can't modify it inside the future, you'd have to construct the oneshot within the function but before the `async` block. Then you can push `rx` into the `self.alive_lend_out_substreams` and move `tx` into the future :)

thomaseizinger reviewed (commented):
The relay protocol is quite complex but you seem to be making great progress which is impressive :)

dgarus committed:
fix review comment

dgarus committed:
Merge branch 'master' into 4075-relay

dgarus committed:
Refactoring of hop protocol

dgarus committed:
Merge branch 'master' into 4075-relay

dgarus committed:
Merge branch 'master' into 4075-relay

dgarus committed:
Merge branch 'master' into 4075-relay

thomaseizinger commented on protocols/relay/src/behaviour/handler.rs:
```diff
@@ -54,28 +61,28 @@ pub struct Config {
 
 pub enum In {
     AcceptReservationReq {
-        inbound_reservation_req: inbound_hop::ReservationReq,
```
Is this necessary? I found the prefixing quite descriptive to quickly see, which protocol this is related to.

thomaseizinger commented on protocols/relay/src/behaviour/handler.rs:
```diff
@@ -380,6 +387,11 @@ pub struct Handler {
+    inbound_requests_futs:
+        FuturesUnordered<BoxFuture<'static, Result<Event, inbound_hop::FatalUpgradeError>>>,
+    stop_requested_streams: VecDeque<StopCommand>,
+    outbound_stop_futs: FuturesUnordered<BoxFuture<'static, Result<Event, FatalUpgradeError>>>,
```
Same here, I'd prefer us to be consistent in prefixing the module for types like `FataUpgradeError`.

thomaseizinger commented on protocols/relay/src/behaviour/handler.rs:
```diff
@@ -396,149 +408,173 @@ impl Handler {
+                Type::STATUS => return Err(inbound_hop::FatalUpgradeError::UnexpectedTypeStatus),
+            };
+
+            Ok(event)
```
What do you think of making this an `async fn` within the `inbound_hop` module? That would remove some of the noise in the this already large `handler.rs` module.

thomaseizinger commented on protocols/relay/src/behaviour/handler.rs:
```diff
@@ -396,149 +408,173 @@ impl Handler {
-            src_peer_id,
-            src_connection_id,
-        } = outbound_open_info;
+    fn send_stop_message_and_process_result(
```
Same here, I think I'd prefer this to be called after the callback (i.e. we do this on every outbound stream).

thomaseizinger commented on protocols/relay/src/behaviour/handler.rs:
```diff
@@ -396,149 +408,173 @@ impl Handler {
-            ));
-    }
+        let mut substream = Framed::new(io, quick_protobuf_codec::Codec::new(MAX_MESSAGE_SIZE));
+        async move {
```
I think this should also be an `async fn` within the `outbound_stop` module.

thomaseizinger commented on protocols/relay/src/behaviour/handler.rs:
```diff
@@ -396,149 +408,173 @@ impl Handler {
+                data: Some(stop_command.max_circuit_bytes),
+            }),
+            status: None,
+        };
```
Move this into the above described function to further reduce the noise in here.

thomaseizinger commented on protocols/relay/src/behaviour/handler.rs:
```diff
@@ -716,6 +739,30 @@ impl ConnectionHandler for Handler {
+                Err(err) => {
+                    ConnectionHandlerEvent::Close(StreamUpgradeError::Apply(Either::Left(err)))
+                }
+            };
```
Instead of pattern matching here, you could directly construct the correct `ConnectionHandlerEvent` within the future, then you could directly `return` the `Output` of the futures here.

thomaseizinger replied:
If you do that, you could even merge both `FuturesUnordered` into one and call them `worker_streams`.

thomaseizinger commented on protocols/relay/src/behaviour/handler.rs:
```diff
@@ -896,33 +943,34 @@ impl ConnectionHandler for Handler {
+                ..
+            }) => {
+                self.inbound_requests_futs
+                    .push(self.process_inbound_request(stream));
```
Instead of having one function that constructs a future and pushes into the `FuturesUnordered` here, I'd prefer to rename `process_inbound_request` to `on_fully_negotiated_inbound` and directly modify `self.inbound_requestes_futs` inside.

thomaseizinger commented on protocols/relay/src/behaviour/handler.rs:
```diff
@@ -931,3 +979,130 @@ pub(crate) struct CircuitParts {
+    max_circuit_bytes: u64,
+}
+
+pub struct ReservationReq {
```
Why move this into here? I'd prefer to keep this file shorter to make it easier to navigate.

dgarus replied:
> I'd prefer to keep this file shorter to make it easier to navigate.

Ok, got it.

> Why move this into here?

Because functions which use them is here ))
If we move these `fn`s to `inbound_hop` then structs would move too.

thomaseizinger replied:
> > Why move this into here?
> 
> Because functions which use them is here ))

Following this logic, everything should be in one file because everything is eventually used by `fn main` :D

Previously, they were in a sub-module of `handler` which I find to be a meaningful structure, i.e. only `handler` needs these structs / functions and only `handler` has access to its own submodules.

dgarus replied:
> Following this logic, everything should be in one file because everything is eventually used by fn main :D

Compiler gathers everything in one file, eventually )))

dgarus replied:
I will fix your comments step-by-step. Thanks!

thomaseizinger reviewed (commented):
Amazing work so far!

I've done a review of the server's handler and the involved protocol. Let me know if you have any questions! Most of the style-suggestions likely apply to the other handler and protocol as well :)

thomaseizinger commented on protocols/relay/src/behaviour/handler.rs:
```diff
@@ -380,6 +387,11 @@ pub struct Handler {
     circuits: Futures<(CircuitId, PeerId, Result<(), std::io::Error>)>,
+
+    inbound_requests_futs:
+        FuturesUnordered<BoxFuture<'static, Result<Event, inbound_hop::FatalUpgradeError>>>,
```
I am working on a `futures-bounded` crate in https://github.com/libp2p/rust-libp2p/pull/3914 that I expect to be merged soon. Once that is the case, I'd like you to adopt it here as well to ensure that these aren't unbounded.

Until then, `FuturesUnbounded` is fine here :)

dgarus replied:
Should I wait for #3914? 
Or we could create a new issue.

dgarus committed:
fix review comments

dgarus committed:
Merge branch 'master' into 4075-relay

dgarus committed:
fix review comments

dgarus committed:
Merge branch 'master' into 4075-relay

dgarus committed:
fix review comments

thomaseizinger commented on protocols/relay/src/behaviour/handler.rs:
```diff
@@ -59,12 +62,12 @@ pub enum In {
     DenyReservationReq {
         inbound_reservation_req: inbound_hop::ReservationReq,
-        status: proto::Status,
+        status: Status,
```
Please try to avoid diffs like these :)

We extensively use the pattern of referencing symbols with 1-2 module names for disambiguation.

thomaseizinger commented on protocols/relay/src/behaviour/handler.rs:
```diff
@@ -18,10 +18,11 @@
-use crate::behaviour::CircuitId;
-use crate::copy_future::CopyFuture;
-use crate::proto;
-use crate::protocol::{inbound_hop, outbound_stop};
```
If possible, it would be great to avoid diffs like these. The `crate::` and `std::` imports seem to just have been flipped but are largely unchanged.

thomaseizinger commented on protocols/relay/src/behaviour/handler.rs:
```diff
@@ -896,33 +779,49 @@ impl ConnectionHandler for Handler {
+                ..
+            }) => {
+                self.inbound_requests_futs.push(
+                    inbound_hop::process_inbound_request(
```
I'd prefer if we establish a convention of what we call the "stream handler" functions within the `inbound_hop`, `outbound_stop` etc modules. Maybe something like `new_stream` or `handle`?

dgarus replied:
`handle_inbound_request`?

thomaseizinger commented on protocols/relay/src/priv_client/handler.rs:
```diff
@@ -18,31 +18,38 @@
 // FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 // DEALINGS IN THE SOFTWARE.
 
-use crate::priv_client::transport;
```
Same here, please keep the diff as minimal as possible.

thomaseizinger commented on protocols/relay/src/priv_client/handler.rs:
```diff
@@ -158,241 +174,303 @@ impl Handler {
+                "Dropping inbound circuit request to be denied from {:?} due to exceeding limit.",
+                src_peer_id
+            );
+        } else if self
```
Please prefer an early return instead of `else if`, i.e. Use `return` in the above block and start with a new `if` after.

thomaseizinger commented on protocols/relay/src/priv_client/handler.rs:
```diff
@@ -158,241 +174,303 @@ impl Handler {
+        }
+    }
+
+    fn send_reserve_message_and_process_response(
```
Should move into the `outbound_hop` module.

thomaseizinger commented on protocols/relay/src/priv_client/handler.rs:
```diff
@@ -402,36 +480,36 @@ impl ConnectionHandler for Handler {
                 dst_peer_id,
             } => {
+                self.wait_for_connection_outbound_stream
+                    .push_back(ConnectionCommand {
```
I think it might work well if we move these "commands" into the respective protocol module. Then we can reference them as `outbound_hop::Command`. What do you think?

thomaseizinger reviewed (commented):
Great progress! I've made some more comments to steer you into the right direction.

Thanks a lot for your help with this PR, it is much appreciated!

dgarus committed:
fix review comments

dgarus committed:
fix review comments

thomaseizinger commented on protocols/relay/src/behaviour/handler.rs:
```diff
@@ -396,149 +402,37 @@ impl Handler {
-        self.pending_error = Some(StreamUpgradeError::Apply(Either::Left(error)));
-    }
-
-    fn on_dial_upgrade_error(
```
I think this function should largely stay as is! We just won't have an `UpgradeError` any more but we should still handle all the other variants.

dgarus replied:
Yes, I agree, but all errors happen inside the `outbound_stop::handle_stop_message_response` method, so I carefully moved errors handling to this method.

thomaseizinger replied:
> but all errors happen inside the `outbound_stop::handle_stop_message_response` method

I don't think that is quite true. `DialUpgradeError` can still return things like `StreamUpgradeError::Timeout` which is currently unhandled. See line 802.

dgarus replied:
> > but all errors happen inside the `outbound_stop::handle_stop_message_response` method
> 
> I don't think that is quite true. `DialUpgradeError` can still return things like `StreamUpgradeError::Timeout` which is currently unhandled. See line 802.

Yes, You are right. Returned the `on_dial_upgrade_error` function.

dgarus committed:
fix review comments

dgarus committed:
fix review comments

dgarus committed:
fix rustfmt

dgarus committed:
fix review comments

dgarus committed:
fix review comments

dgarus committed:
fix review comments

dgarus committed:
ConnectionCommand to Command

dgarus committed:
Merge branch 'master' into 4075-relay

dgarus committed:
rename `fn`s

dgarus committed:
fix rustfmt

dgarus committed:
Merge branch 'master' into 4075-relay

thomaseizinger commented on protocols/relay/src/priv_client/handler.rs:
```diff
@@ -140,8 +151,11 @@ pub struct Handler {
-        HashMap<PeerId, BoxFuture<'static, Result<(), protocol::inbound_stop::UpgradeError>>>,
+    open_circuit_futs: FuturesUnordered<
+        BoxFuture<'static, Result<inbound_stop::Circuit, inbound_stop::FatalUpgradeError>>,
+    >,
```
It would be great if we could unify these `FuturesUnordered` somehow.

dgarus replied:
'FuturesUnordered' is lesser evil compared to boilerplate code related to `VecDeque<SomeState>`. 
IMHO, using `OutboundOpenInfo`/`InboundOpenInfo` looks much more convenient.

mxinden replied:
I see the additional boilerplate. Thanks for pointing it out. 

That said, I don't think `OutboundOpenInfo` and InboundOpenInfo` are worth the complexity they introduce. Thus long term I would really like to get rid of them.

thomaseizinger reviewed (commented):
This looks great! I'll give this a more in-depth review later :)

dgarus committed:
fix review comments

dgarus committed:
Merge branch 'master' into 4075-relay

dgarus committed:
Merge branch 'master' into 4075-relay

dgarus committed:
added futures-bounded struct

dgarus committed:
added futures-bounded struct

dgarus committed:
Merge branch 'master' into 4075-relay

mxinden commented on protocols/autonat/src/behaviour.rs:
This change should be removed.

mxinden commented on protocols/perf/src/client.rs:
This change should be removed.

mxinden commented on protocols/perf/src/server.rs:
This change should be removed.

mxinden commented on protocols/relay/src/behaviour/handler.rs:
```diff
@@ -549,26 +482,26 @@ enum ReservationRequestFuture {
-    type InboundProtocol = inbound_hop::Upgrade;
-    type OutboundProtocol = outbound_stop::Upgrade;
-    type OutboundOpenInfo = OutboundOpenInfo;
+    type InboundProtocol = ReadyUpgrade<StreamProtocol>;
     type InboundOpenInfo = ();
+    type OutboundProtocol = ReadyUpgrade<StreamProtocol>;
+    type OutboundOpenInfo = ();
```
:+1:

mxinden commented on protocols/relay/src/priv_client/handler.rs:
```diff
@@ -128,6 +129,17 @@ pub struct Handler {
+    reserve_futs: futures_bounded::WorkerFutures<
+        (),
+        Result<outbound_hop::Output, outbound_hop::UpgradeError>,
+    >,
```
Definition of `WorkerFutures` looks right.

dgarus replied:
@mxinden I wonder about
```
reserve_futs: futures_bounded::WorkerFutures::new(
                config.substream_timeout,
                config.max_concurrent_streams,
            ),
            
```

Is it a right way, or I should define `substream_timeout` and `max_concurrent_streams`, for example, as constants?

mxinden replied:
I don't have a strong opinion on whether `substream_timeout` and `max_concurrent_streams` should be configurable. You @thomaseizinger. For the sake of keeping the API surface small, I am fine settling on constants for now. In case we ever want to configure them, we can do so in a non-breaking way.

dgarus replied:
I'll change to constants usage

mxinden commented on protocols/request-response/CHANGELOG.md:
This change should be removed.

mxinden commented on protocols/request-response/Cargo.toml:
This change should be removed.

mxinden commented on protocols/request-response/src/handler.rs:
This change should be removed.

mxinden commented on protocols/request-response/src/handler/protocol.rs:
This change should be removed.

mxinden commented on protocols/request-response/src/lib.rs:
This change should be removed.

mxinden reviewed (commented):
As always, your work here is much appreciated.

dgarus committed:
Removed extra changes

dgarus committed:
Removed extra changes

dgarus committed:
Merge branch 'master' into 4075-relay

dgarus committed:
using futures-bounded struct

dgarus committed:
using futures-bounded struct

dgarus committed:
Merge branch 'master' into 4075-relay

dgarus committed:
using futures-bounded struct

dgarus committed:
fix format

dgarus committed:
Merge branch 'master' into 4075-relay

dgarus committed:
Merge branch 'master' into 4075-relay

mxinden commented on protocols/relay/src/behaviour.rs:
```diff
@@ -54,6 +54,8 @@ pub struct Config {
     pub reservation_rate_limiters: Vec<Box<dyn rate_limiter::RateLimiter>>,
 
+    pub substream_timeout: Duration,
+    pub max_concurrent_streams: usize,
```
```suggestion
    pub max_concurrent_streams_per_connection: usize,
```

Would be more descriptive, right?

mxinden commented on protocols/relay/src/behaviour.rs:
```diff
@@ -151,6 +153,8 @@ impl Default for Config {
             reservation_rate_limiters,
 
+            substream_timeout: Duration::from_secs(10),
+            max_concurrent_streams: 100,
```
Is there some rational for choosing `100` here? This is the max per connection, correct? Why have two orders of magnitude more streams than the allowed number of circuits per peer? `10` seems more reasonable to me. Would you agree?

```suggestion
            max_concurrent_streams: 10,
```

dgarus replied:
I copied `100` from Tomas's PR, no deep thinking ))

mxinden commented on protocols/relay/src/behaviour/handler.rs:
```diff
@@ -549,26 +502,26 @@ enum ReservationRequestFuture {
 
 type Futures<T> = FuturesUnordered<BoxFuture<'static, T>>;
 
+pub(crate) type CHEvent = ConnectionHandlerEvent<
```
We don't abbreciate `ConnectionHandler` as `CH` anywhere else in the codebase. How about the following?

```suggestion
pub(crate) type ConnectionHandlerEvent = libp2p_swarm::ConnectionHandlerEvent<
```

Would just require the removal of the `use libp2p_swarm::ConnectionHandler` above.

dgarus replied:
Renamed to 'RelayConnectionHandlerEvent' otherwise it is confusing with `ConnectionHandlerEvent` from `libp2p_swarm` in the code.

mxinden commented on protocols/relay/src/behaviour.rs:
```diff
@@ -54,6 +54,8 @@ pub struct Config {
     pub reservation_duration: Duration,
     pub reservation_rate_limiters: Vec<Box<dyn rate_limiter::RateLimiter>>,
 
+    pub substream_timeout: Duration,
```
```suggestion
    pub stream_timeout: Duration,
```

We are moving away from the **sub**stream term, and instead use **stream** everywhere. See also https://github.com/libp2p/rust-libp2p/pull/3912.

mxinden commented on protocols/relay/src/behaviour/handler.rs:
```diff
@@ -716,6 +665,14 @@ impl ConnectionHandler for Handler {
         }
 
+        // Process protocol requests
+        if let Poll::Ready((_, worker_res)) = self.protocol_futs.poll_unpin(cx) {
```
```suggestion
        if let Poll::Ready(((), worker_res)) = self.protocol_futs.poll_unpin(cx) {
```

This is the ID, right? In case we ever change it away from `()` we get a helpful compiler error here.

mxinden commented on protocols/relay/src/behaviour/handler.rs:
```diff
@@ -380,11 +382,18 @@ pub struct Handler {
     circuits: Futures<(CircuitId, PeerId, Result<(), std::io::Error>)>,
+
+    stop_requested_streams: VecDeque<outbound_stop::StopCommand>,
+    protocol_futs: futures_bounded::WorkerFutures<(), CHEvent>,
```
I don't find this property name intuitive. That said I can not think of a better name. I am fine merging with the current name.

thomaseizinger replied:
How about `workers` or `worker_streams`?

mxinden commented on protocols/relay/src/priv_client.rs:
```diff
@@ -88,9 +89,26 @@ pub enum Event {
 
+#[derive(Debug, Clone)]
+pub struct Config {
+    pub substream_timeout: Duration,
```
```suggestion
    pub stream_timeout: Duration,
```

mxinden commented on protocols/relay/src/priv_client.rs:
```diff
@@ -88,9 +89,26 @@ pub enum Event {
+#[derive(Debug, Clone)]
+pub struct Config {
+    pub substream_timeout: Duration,
+    pub max_concurrent_streams: usize,
```
```suggestion
    pub max_concurrent_streams_per_connection: usize,
```

mxinden commented on protocols/relay/src/priv_client/handler.rs:
```diff
@@ -140,8 +153,12 @@ pub struct Handler {
+        Result<inbound_stop::Circuit, inbound_stop::FatalUpgradeError>,
+    >,
+
+    circuit_deny_futs: HashMap<PeerId, BoxFuture<'static, Result<(), inbound_stop::UpgradeError>>>,
```
How about making this a `futures_bounded` as well? Would implicitly enforce `MAX_NUMBER_DENYING_CIRCUIT`. In other words simplify the code below.

dgarus replied:
@mxinden @thomaseizinger 
Good point.
But there are a few questions:
1. We should add to `futures_bounded` ability to remember which it has keys, isn't?
 Otherwise, it changes functionality, we could add the same key twice.

2. What should we do when a timeout happens?

thomaseizinger replied:
> * We should add to `futures_bounded` ability to remember which it has keys, isn't?
>   Otherwise, it changes functionality, we could add the same key twice.

Overall, that would make sense but the problem is that `()` is always equal to `()`. Thus, if we enforce that `futures-bounded` can't have duplicate keys, then we can't use it with `()` as a key.

We'd have to special-case that in `futures-bounded` by creating separate `impl` blocks with functions that don't allow you to pass a worker ID and thus also don't enforce its uniqueness. I am not sure that is worth the effort?

> * What should we do when a timeout happens?

We can just log an error. At the moment, we don't have any timeout actually! This shows how important this PR is :)

mxinden commented on protocols/relay/src/priv_client/handler.rs:
```diff
@@ -458,13 +357,139 @@ impl ConnectionHandler for Handler {
+        }
+
+        // Circuit connections
+        if let Poll::Ready((_, worker_res)) = self.circuit_connection_futs.poll_unpin(cx) {
```
I consider reservations a source of new circuit connections. With that in mind, I suggest first polling existing connections (i.e. `self.circuit_connection_futs`) before polling the thing that leads to new circuit connections (i.e. `self.reserve_futs`). 

See also https://github.com/libp2p/rust-libp2p/blob/master/docs/coding-guidelines.md#prioritize-local-work-over-new-work-from-a-remote

Does that make sense here?

dgarus replied:
Yes, good point!

mxinden commented on protocols/relay/src/priv_client/handler.rs:
```diff
@@ -458,13 +357,139 @@ impl ConnectionHandler for Handler {
+                return Poll::Ready(ConnectionHandlerEvent::Close(StreamUpgradeError::Timeout));
+            }
+
+            let opt = match worker_res.unwrap() {
```
Instead of the `unwrap`, would a `match worker_res` work here?

```suggestion
            let opt = match worker_res {
                Err(the-concrete-type-here) => {
                    return Poll::Ready(ConnectionHandlerEvent::Close(StreamUpgradeError::Timeout));
                },
                Ok(opt) => opt,
            };
```

dgarus replied:
Sure. 
Initially, I did like you suggested, but thought that this way looks a little shorter.

mxinden commented on protocols/relay/src/protocol/outbound_hop.rs:
In my eyes this pull request significantly reduces the complexity in `relay/src/protocol`. Good job!

dgarus replied:
Thanks, glad to hear this!

mxinden reviewed (commented):
Well done working yourself through this difficult code-base!

dgarus committed:
Update protocols/relay/src/behaviour.rs

Co-authored-by: Max Inden <mail@max-inden.de>

dgarus committed:
rename `max_concurrent_streams` to `max_concurrent_streams_per_connection`

dgarus committed:
Update protocols/relay/src/behaviour.rs

Co-authored-by: Max Inden <mail@max-inden.de>

dgarus committed:
Update protocols/relay/src/behaviour/handler.rs

Co-authored-by: Max Inden <mail@max-inden.de>

dgarus committed:
Merge remote-tracking branch 'origin/4075-relay' into 4075-relay

# Conflicts:
#	protocols/relay/src/behaviour.rs

dgarus committed:
fix review comments

dgarus committed:
removed config fields

dgarus committed:
Update protocols/relay/src/behaviour/handler.rs

Co-authored-by: Max Inden <mail@max-inden.de>

dgarus committed:
fix review comments

dgarus committed:
Merge branch 'master' into 4075-relay

# Conflicts:
#	Cargo.toml

dgarus committed:
fix toml file

dgarus committed:
fix review comments

dgarus committed:
fix review comments

dgarus committed:
refactoring

dgarus committed:
аix failed tests

dgarus committed:
Merge branch 'master' into 4075-relay

dgarus committed:
Merge branch 'master' into 4075-relay

thomaseizinger commented on protocols/relay/src/protocol/outbound_stop.rs:
```diff
@@ -161,3 +73,152 @@ pub enum FatalUpgradeError {
+
+    match status {
+        Some(proto_status) => match proto_status {
+            proto::Status::OK => {}
```
These `match` can be nested which will reduce the level of indentation here which makes this easier to read.

```suggestion
    match status {
        Some(proto::Status::OK) => {}
```

thomaseizinger commented on protocols/relay/src/protocol/outbound_stop.rs:
```diff
@@ -161,3 +73,152 @@ pub enum FatalUpgradeError {
+    let mut substream = Framed::new(io, quick_protobuf_codec::Codec::new(MAX_MESSAGE_SIZE));
+
+    let send_res = substream.send(msg).await;
+    if send_res.is_err() {
```
Do we (still) need this split where we have to assign the result or can we just inline this variable?

```suggestion
    if substream.send(msg).await.is_err() {
```

thomaseizinger commented on protocols/relay/src/behaviour/handler.rs:
```diff
@@ -30,21 +30,24 @@ use futures::io::AsyncWriteExt;
 use std::task::{Context, Poll};
 use std::time::Duration;
 
+const STREAM_TIMEOUT: Duration = Duration::from_secs(60 * 60);
```
An hour is a bit long, don't you think? 60 seconds seems to be more appropriate.

dgarus replied:
Yes, ok

thomaseizinger commented on protocols/relay/src/behaviour/handler.rs:
```diff
@@ -380,11 +383,18 @@ pub struct Handler {
     /// Futures relaying data for circuit between two peers.
     circuits: Futures<(CircuitId, PeerId, Result<(), std::io::Error>)>,
+
+    stop_requested_streams: VecDeque<outbound_stop::StopCommand>,
```
To expand on https://github.com/libp2p/rust-libp2p/issues/3268#issuecomment-1694381620, I was thinking of something like:

```rust
struct OutboundStreams<TInfo> {
	info: VecDeque<TInfo>,
	open_stream_results: VecDeque<Result<Stream, StreamUpgradeError<()>>>,

	// possibly more state ...
}

impl<TInfo> OutboundStreams<TInfo> {
	pub fn request_new_stream(&mut self, info: TInfo);

	pub fn on_new_stream(&mut self, stream: Stream);
	pub fn on_stream_error(&mut self, error: StreamUpgradeError<()>);
	
	pub fn poll(&mut self, cx: Context<'_>) -> Poll<Event<TInfo>>>;
}

enum Event<TInfo> {
	
	Stream(Result<(Stream, TInfo), Error<TInfo>),
	/// Instruct the `ConnectionHandler` to request a new stream.
	RequestNewStream
}

enum Error<TInfo> {
	Timeout(TInfo),
    	Unsupported(TInfo),
	Io((TInfo, io::Error)),
}
```

The user would store one of these in their `ConnectionHandler` in place of the above `VecDeque<outbound_stop::StopCommand>`. You then call the `on_new_stream` function in the `on_connection_event` callback and `poll` inside the `ConnectionHandler::poll`.

Have a play with the design and let me know what you think :)

dgarus replied:
It looks interesting.
I think this is a must-have helping tool, and we should implement something like this. But let's do this in a separate issue. Otherwise, the current issue will go far behind the scope of the relay feature.
I'm ready to make this tool right after the closing of this PR.
WDYT?

thomaseizinger replied:
Sounds good! Thank you for your help :)

thomaseizinger replied:
Resolving this because otherwise we can't merge.

thomaseizinger replied:
I've recorded this in a follow-up issue: https://github.com/libp2p/rust-libp2p/issues/4510

dgarus committed:
Merge branch 'master' into 4075-relay

dgarus committed:
Merge branch 'master' into 4075-relay

dgarus committed:
Merge branch 'master' into 4075-relay

dgarus committed:
Merge branch 'master' into 4075-relay

dgarus committed:
Merge branch 'master' into 4075-relay

dgarus committed:
fix review comments

dgarus committed:
bounded and unique workers sets

dgarus committed:
Merge branch 'master' into 4075-relay

dgarus committed:
Merge branch 'master' into 4075-relay

dgarus committed:
Merge branch 'master' into 4075-relay

thomaseizinger commented on misc/futures-bounded/src/lib.rs:
```diff
@@ -103,91 +15,20 @@ impl Timeout {
-    }
+/// Result of a worker pushing
+#[derive(PartialEq, Debug)]
+pub enum PushResult {
```
I'd like this to reuse `Result` and have a dedicated error instead.

thomaseizinger commented on misc/futures-bounded/src/unique.rs:
```diff
@@ -0,0 +1,218 @@
+    {
+        // Check worker's ID
+        if !self.keys.insert(worker_id.clone()) {
+            return PushResult::ExistedID;
```
Should we return the worker instead? I think it would be a better API if we leave it up to the user to drop it if they don't need it.

dgarus replied:
ok

thomaseizinger commented on misc/futures-bounded/src/bounded.rs:
```diff
@@ -0,0 +1,186 @@
+    /// This method adds the given worker to the set and returns [PushResult::Ok].
+    /// If length of the set is equal the capacity, this method returns [PushResult::BeyondCapacity]
+    /// and worker is ignored.
+    pub fn try_push<F>(&mut self, worker: F) -> PushResult
```
Ideally, this API doesn't use `PushResult` because it will never return the 2nd error regarding IDs. We shouldn't make users handle errors that will never occur :)

thomaseizinger reviewed (commented):
Some ideas!

dgarus committed:
rewrite unique workers

dgarus committed:
Merge branch 'master' into 4075-relay

dgarus committed:
fix docs

dgarus committed:
fix clippy

dgarus committed:
rewrite unique workers

dgarus committed:
Merge branch 'master' into 4075-relay

dgarus committed:
wrapping UniqueWorkers

dgarus committed:
rustfmt

dgarus committed:
fix doc

thomaseizinger commented on misc/futures-bounded/src/bounded.rs:
```diff
@@ -40,64 +29,40 @@ impl<O> BoundedWorkers<O> {
-
-                Poll::Ready(result)
-            }
+        match self.inner.poll_unpin(cx) {
```
You can use the `futures::ready!` macro here :)

thomaseizinger commented on misc/futures-bounded/src/unique.rs:
```diff
@@ -0,0 +1,230 @@
+
+        match res {
+            None => {
+                self.empty_waker = Some(cx.waker().clone());
```
This is not necessarily true, right? We also return `None` if all workers are `Pending` in which case you would be allowed to just return `Poll::Pending` without registering a waker yourself.

I'd suggest that we return `Option<Poll>` from the `find_map` closure, then you can correctly differentiate between all these cases. `Poll::Pending` would have to be translated to `Some(Poll::Pending)`.

dgarus replied:
Good point!
I suggest making an explicit check for empty map.

dgarus replied:
I've been thinking, there are no tests to prove the correctness of working `empty_waker` and `full_waker`. 
It looks wrong, isn't?

thomaseizinger replied:
I pushed a commit that uses `FuturesUnordered` under the hood which should be more efficient because it polls only the futures that actually need to be polled.

thomaseizinger commented on misc/futures-bounded/src/unique.rs:
```diff
@@ -0,0 +1,230 @@
+    ID: Clone + Hash + Eq + Send + 'static,
+{
+    /// Push a worker into the set.
+    /// This method adds the given worker with defined `worker_id` to the set.
```
```suggestion
    /// Push a worker into the set.
    ///
    /// This method adds the given worker with defined `worker_id` to the set.
```

By convention, rustdoc has a newline here :)

thomaseizinger commented on misc/futures-bounded/src/unique.rs:
```diff
@@ -0,0 +1,230 @@
+    /// This method adds the given worker with defined `worker_id` to the set.
+    /// If the length of the set is equal to the capacity, this method returns [PushError::BeyondCapacity],
+    /// that contains the passed worker. In that case, the worker is not added to the set.
+    /// If a worker with the given `worker_id` already exists, then the old worker will be replaced by a new one.
```
This is opinionated but I think it is the behaviour that makes the most sense :)

thomaseizinger commented on misc/futures-bounded/src/bounded.rs:
```diff
@@ -0,0 +1,152 @@
+/// Each worker must finish within the specified time and the set never outgrows its capacity.
+pub struct BoundedWorkers<O> {
+    id: i32,
+    inner: UniqueWorkers<i32, O>,
```
As I was reviewing this, I was constantly confused between which of these two structs is which. I'd like to rename them.

`futures` has set a precedent with the `FuturesUnordered` and `FuturesOrdered` types. I think we should follow the same naming scheme. Maybe we can call them `WorkerMap` and `WorkerList`? Or maybe we call them `FuturesMap` and `FuturesList` and move away from the "worker" terminology?

thomaseizinger commented on protocols/relay/src/behaviour/handler.rs:
```diff
@@ -396,93 +406,51 @@ impl Handler {
+                    self.endpoint.clone(),
+                    self.active_reservation.is_some(),
+                )
+                .boxed(),
```
Shouldn't need the boxing here, right? The library does it for us :)

thomaseizinger commented on protocols/relay/src/behaviour/handler.rs:
```diff
@@ -396,93 +406,51 @@ impl Handler {
-        self.pending_error = Some(StreamUpgradeError::Apply(Either::Left(error)));
+        if self
+            .protocol_futs
+            .try_push(outbound_stop::handle_stop_message_response(stream, stop_command, tx).boxed())
```
```suggestion
            .try_push(outbound_stop::handle_stop_message_response(stream, stop_command, tx))
```

Same here.

thomaseizinger commented on protocols/relay/src/behaviour/handler.rs:
```diff
@@ -502,39 +470,21 @@ impl Handler {
-                    (StreamUpgradeError::Apply(error), status)
-                }
-            },
+            StreamUpgradeError::Apply(_) => unreachable!("Should not emit handle errors"),
```
The handler error should be a `Void` so you can do:

```suggestion
            StreamUpgradeError::Apply(v) => void::unreachable(v),
```

thomaseizinger commented on protocols/relay/src/behaviour/handler.rs:
```diff
@@ -716,6 +662,14 @@ impl ConnectionHandler for Handler {
+        // Process protocol requests
+        if let Poll::Ready(worker_res) = self.protocol_futs.poll_unpin(cx) {
+            let event = worker_res
+                .unwrap_or_else(|_| ConnectionHandlerEvent::Close(StreamUpgradeError::Timeout));
```
Ideally, we shouldn't be closing the connection here but that is a separate issue (https://github.com/libp2p/rust-libp2p/issues/3591). This just mirrors the existing behaviour as far as I can understand.

thomaseizinger commented on protocols/relay/src/priv_client/handler.rs:
```diff
@@ -158,183 +176,50 @@ impl Handler {
-                            unreachable!("Do not emitt `CircuitFailed` for outgoing reservation.")
-                        }
-                    },
+                    StreamUpgradeError::Apply(_) => unreachable!("should not update"),
```
Same here, I think this can use `void::unreachable`.

```suggestion
                    StreamUpgradeError::Apply(v) => void::unreachable(v),
```

thomaseizinger commented on protocols/relay/src/priv_client/handler.rs:
```diff
@@ -360,30 +246,18 @@ impl Handler {
-                            unreachable!("Do not emitt `ReservationFailed` for outgoing circuit.")
-                        }
-                    },
+                    StreamUpgradeError::Apply(_) => unreachable!("should not update"),
```
```suggestion
                    StreamUpgradeError::Apply(v) => void::unreachable(v),
```

thomaseizinger commented on protocols/relay/src/priv_client/handler.rs:
```diff
@@ -458,38 +351,167 @@ impl ConnectionHandler for Handler {
+                }
+            };
+
+            let opt = match res {
```
Instead of assigning to `opt` here, I think I'd prefer just using `return` within the branches were we need to.

thomaseizinger commented on protocols/relay/src/priv_client/handler.rs:
```diff
@@ -458,38 +351,167 @@ impl ConnectionHandler for Handler {
+                }
+                Ok(None) => None,
+                Err(err) => {
+                    let res = match err {
```
Can collapse these `match`es.

```suggestion
                Err(outbound_hop::UpgradeError::CircuitFailed(e)) => {
```

thomaseizinger commented on protocols/relay/src/priv_client/handler.rs:
```diff
@@ -458,38 +351,167 @@ impl ConnectionHandler for Handler {
+                            StreamUpgradeError::Apply(Either::Right(e)),
+                        ),
+                        outbound_hop::UpgradeError::ReservationFailed(_) => {
+                            unreachable!("do not emit `ReservationFailed` for connection")
```
I am not very fond of having these `unreachable!` calls here. What do you think of just changing the API signature of `outbound_hop::handle_reserve_message_response` to return a `Result<Option>` by wrapping the return value in `Some`. Then we could merge these to `BoundedWorkers` and we wouldn't have these panics here.

WDYT?

dgarus replied:
I think this is an excellent idea.

thomaseizinger commented on protocols/relay/src/priv_client/handler.rs:
```diff
@@ -533,19 +555,65 @@ impl ConnectionHandler for Handler {
+            }) => {
+                if self
+                    .open_circuit_futs
+                    .try_push(inbound_stop::handle_open_circuit(stream).boxed())
```
```suggestion
                    .try_push(inbound_stop::handle_open_circuit(stream))
```

thomaseizinger commented on protocols/relay/src/priv_client/handler.rs:
```diff
@@ -533,19 +555,65 @@ impl ConnectionHandler for Handler {
+                            .reserve_futs
+                            .try_push(
+                                outbound_hop::handle_reserve_message_response(stream, to_listener)
+                                    .boxed(),
```
```suggestion
```

thomaseizinger commented on protocols/relay/src/priv_client/handler.rs:
```diff
@@ -533,19 +555,65 @@ impl ConnectionHandler for Handler {
+                                    cmd,
+                                    tx,
+                                )
+                                .boxed(),
```
```suggestion
```

thomaseizinger reviewed (commented):
This looks very clean already. A few more smaller tidy-ups but I think we can merge this soon :tada: 

Thank you for this prolonged and massive effort! It is much appreciated :)

dgarus committed:
fix review comments

dgarus committed:
Update misc/futures-bounded/src/unique.rs

Co-authored-by: Thomas Eizinger <thomas@eizinger.io>

dgarus committed:
Update protocols/relay/src/priv_client/handler.rs

Co-authored-by: Thomas Eizinger <thomas@eizinger.io>

dgarus committed:
Update protocols/relay/src/priv_client/handler.rs

Co-authored-by: Thomas Eizinger <thomas@eizinger.io>

dgarus committed:
Update protocols/relay/src/priv_client/handler.rs

Co-authored-by: Thomas Eizinger <thomas@eizinger.io>

dgarus committed:
Update protocols/relay/src/priv_client/handler.rs

Co-authored-by: Thomas Eizinger <thomas@eizinger.io>

dgarus committed:
Merge branch 'master' into 4075-relay

dgarus committed:
fix review comments

dgarus committed:
Renamed to FuturesList and FuturesMap

dgarus committed:
fix review comments

dgarus committed:
refactoring of outbound_hop::handle_reserve_message_response

dgarus committed:
refactoring of outbound_hop::handle_reserve_message_response

dgarus committed:
Merge branch 'master' into 4075-relay

thomaseizinger committed:
Refactor `FuturesMap` to use `FuturesUnbounded` for efficiency

thomaseizinger committed:
Remove tests that are covered by other impl

Now that we are implementing `FuturesList` by wrapping `FuturesMap`, we
no longer need these tests for `FuturesList`.

thomaseizinger committed:
Reformat docs

thomaseizinger committed:
Improve panic message

thomaseizinger committed:
Make `Timeout` a proper error

thomaseizinger committed:
Handle outbound circuits using 1 `FuturesList`

thomaseizinger committed:
Don't nest `match`

thomaseizinger committed:
Don't publish `futures-bounded` for now

thomaseizinger committed:
Inline type-alias that is only used once

thomaseizinger committed:
Use `void::unreachable`

thomaseizinger committed:
Log PeerIds using Display

thomaseizinger committed:
Use `void::unreachable`

thomaseizinger committed:
Reduce diff

thomaseizinger committed:
Remove commented out code

thomaseizinger reviewed (approved):
Thank you so much for this big contribution @dgarus!

As a final request, because it is such a large change. Do you mind doing some manual testing of the various examples to check that they still work?

We don't have automated, cross-implementation circuit-relay tests yet unfortunately :/

thomaseizinger committed:
Use `void::unreachable`

thomaseizinger committed:
Rename function to resemble message that is being sent

thomaseizinger committed:
Rename `StopCommand` to `PendingConnect`

thomaseizinger committed:
Remove circular dependencies

dgarus commented on protocols/relay/src/behaviour/handler.rs:
```diff
@@ -660,11 +668,63 @@ impl ConnectionHandler for Handler {
-
-            return Poll::Ready(event);
+        match self.workers.poll_unpin(cx) {
+            Poll::Ready(Ok(Either::Left(Ok(Either::Left(inbound_reservation_req))))) => {
```
@thomaseizinger
`Poll::Ready(Ok(Either::Left(Ok(Either::Left(inbound_reservation_req)))))`
IMHO, It is hard to keep in mind what means every Either::Left.
Maybe it would be more convenient to create a certain `enum` for every `Either`?

thomaseizinger replied:
Yeah we could! To be honest, I mostly rely on the compiler here to make sure I don't miss any combinations. Personally, I am not too worried about the complexity of these matches because it is very hard to make a mistake :)

If you feel strongly, feel free to make a custom enum but I am not sure it is worth the code addition.

dgarus replied:
>  I am not sure it is worth the code addition.

ok

dgarus committed:
Merge branch 'master' into 4075-relay

dgarus committed:
Merge branch 'master' into 4075-relay

dgarus committed:
Merge branch 'master' into 4075-relay

thomaseizinger committed:
Merge branch 'master' into 4075-relay

thomaseizinger committed:
Fix clippy lint

thomaseizinger committed:
Allow beta lint

thomaseizinger committed:
Merge branch 'master' into 4075-relay

dgarus committed:
Merge branch 'master' into 4075-relay

dgarus committed:
Merge branch 'master' into 4075-relay

mxinden commented on misc/futures-bounded/src/list.rs:
```diff
@@ -0,0 +1,58 @@
+    where
+        F: Future<Output = O> + Send + 'static,
+    {
+        (self.id, _) = self.id.overflowing_add(1);
```
```suggestion
        (self.id, _) = self.id.checked_add(1).expect("i32 to never be exhausted");
```

We expect this to never overflow, correct? Overflowing would panic further below, correct? If so, let's panic here already.

dgarus replied:
No, 'id' can overflow. 

> Overflowing would panic further below, correct?

No, we get a negative value for `id`.
We just need a unique `id` here, so if it would be negative it's ok.

dgarus replied:
But I can make this a simpler way.

mxinden commented on misc/futures-bounded/src/list.rs:
```diff
@@ -0,0 +1,58 @@
+/// Represents a list of [Future]s.
+///
+/// Each future must finish within the specified time and the list never outgrows its capacity.
+pub struct FuturesList<O> {
```
Given that the "list" does not expose its order, it is rather a "set", no?

mxinden commented on misc/futures-bounded/src/list.rs:
```diff
@@ -0,0 +1,58 @@
+///
+/// Each future must finish within the specified time and the list never outgrows its capacity.
+pub struct FuturesList<O> {
+    id: i32,
```
Why the complexity of a signed integer? Why not a `u32`?

mxinden commented on misc/futures-bounded/src/map.rs:
```diff
@@ -0,0 +1,268 @@
+
+use crate::Timeout;
+
+/// Represents a map of [Future]s.
```
```suggestion
/// Represents a map of [`Future`]s.
```

mxinden reviewed (commented):
Just some small nit-picks. Otherwise this looks good to me. Thank you for the work!

dgarus committed:
fix review comments

dgarus committed:
Update misc/futures-bounded/src/map.rs

Co-authored-by: Max Inden <mail@max-inden.de>

startup-dreamer committed:
feat(swarm): allow configuration to idle connection timeout

Previously, a connection would be shut down immediately as soon as its `ConnectionHandler` reports `KeepAlive::No`. As we have gained experience with libp2p, it turned out that this isn't ideal.

For one, tests often need to keep connections alive longer than the configured protocols require. Plus, some usecases require connections to be kept alive in general.

Both of these needs are currently served by the `keep_alive::Behaviour`. That one does essentially nothing other than statically returning `KeepAlive::Yes` from its `ConnectionHandler`.

It makes much more sense to deprecate `keep_alive::Behaviour` and instead allow users to globally configure an `idle_conncetion_timeout` on the `Swarm`. This timeout comes into effect once a `ConnectionHandler` reports `KeepAlive::No`. To start with, this timeout is 0. Together with https://github.com/libp2p/rust-libp2p/issues/3844, this will allow us to move towards a much more aggressive closing of idle connections, together with a more ergonomic way of opting out of this behaviour.

Fixes #4121.

Pull-Request: #4161.

thomaseizinger committed:
Fix formatting

dgarus committed:
Merge branch 'master' into 4075-relay
